rasa_core/featurizers.py

```diff
@@ -161,20 +157,443 @@ def encode(self, active_features, input_feature_map):
+      The flag that specifies if to create the same vocabulary for
+      user intents and bot actions."""
+
+    def __init__(self, split_symbol='_', share_vocab=False):
+        # type: (Text, bool) -> None
+        """inits vocabulary for label bag of words representation"""
+
+        self.share_vocab = share_vocab
+        self.split_symbol = split_symbol
+
+        self.num_features = None
+        self.user_labels = []
+        self.bot_labels = []
+        self.other_labels = []
+
+        self.bot_vocab = None
+        self.user_vocab = None
+
+    @staticmethod
+    def _create_label_token_dict(labels, split_symbol='_'):
+        """Splits labels into tokens by using provided symbol.
+        Creates the lookup dictionary for this tokens.
+        Values in this dict are used for featurization."""
+
+        distinct_tokens = set([token
+                               for label in labels
+                               for token in label.split(split_symbol)])
+        return {token: idx
+                for idx, token in enumerate(sorted(distinct_tokens))}
+
+    def prepare_from_domain(self, domain):
+        # type: (Domain) -> None
+        """Creates internal vocabularies for user intents
+        and bot actions to use for featurization"""
+        self.user_labels = domain.intent_states + domain.entity_states
+        self.bot_labels = domain.action_names
+        self.other_labels = domain.slot_states
+
+        if self.share_vocab:
+            self.bot_vocab = self._create_label_token_dict(self.bot_labels +
+                                                           self.user_labels,
+                                                           self.split_symbol)
+            self.user_vocab = self.bot_vocab
+        else:
+            self.bot_vocab = self._create_label_token_dict(self.bot_labels,
+                                                           self.split_symbol)
+            self.user_vocab = self._create_label_token_dict(self.user_labels,
+                                                            self.split_symbol)
+
+        self.num_features = (len(self.user_vocab) +
+                             len(self.bot_vocab) +
+                             len(self.other_labels))
+
+    def encode(self, states):
+        # type: (Optional[Text, float]) -> np.ndarray
+        if not self.num_features:
+            raise Exception("LabelTokenizerSingleStateFeaturizer "
+                            "was not prepared before encoding.")
+
+        if states is None or None in states:
+            return np.ones(self.num_features, dtype=int) * -1
+
+        used_features = np.zeros(self.num_features, dtype=int)
+        for state_name, prob in states.items():
+
+            if state_name in self.user_labels:
+                if PREV_PREFIX + ACTION_LISTEN_NAME not in states:
+                    # we predict next action from bot action
+                    # TODO do we need state_name = 'intent_listen' ?
```

Comment:
since I deleted previous intent from feature list, if user did not input anything, I was thinking whether it'd make sense to create `intent_listen` instead of zero feature vector

Reply:
ah ok, so when the conversation is started and the state list is still completely empty?

Reply:
yes, but I think that is not needed, I'll leave this TODO for now as a reminder, we'll remove it before merge

Reply:
@Ghostvv so should we remove this, wdyt?